In [8]:
import pandas as pd
import numpy as np

In [3]:
df = pd.read_csv("BL-Flickr-Images-Book.csv")
df.head(3)

Identifier  ...                         Shelfmarks
0         206  ...  British Library HMNTS 12641.b.30.
1         216  ...  British Library HMNTS 12626.cc.2.
2         218  ...  British Library HMNTS 12625.dd.1.

[3 rows x 15 columns]

In [4]:
df.columns

Index(['Identifier', 'Edition Statement', 'Place of Publication',
       'Date of Publication', 'Publisher', 'Title', 'Author', 'Contributors',
       'Corporate Author', 'Corporate Contributors', 'Former owner',
       'Engraver', 'Issuance type', 'Flickr URL', 'Shelfmarks'],
      dtype='object')

In [9]:
print("Shape of the dataframe: {}".format(df.shape))

Shape of the dataframe: (8287, 15)


Let's see all the missing values for each column

In [6]:
df.isnull().sum()

Identifier                   0
Edition Statement         7514
Place of Publication         0
Date of Publication        181
Publisher                 4195
Title                        0
Author                    1778
Contributors                 0
Corporate Author          8287
Corporate Contributors    8287
Former owner              8286
Engraver                  8287
Issuance type                0
Flickr URL                   0
Shelfmarks                   0
dtype: int64

We look there are columns that all the values have missing data and we can see that a handful of columns provide ancillary information that would be helpful to the library but is not very descriptive of the book themselves

In [13]:
drop_columns = ["Edition Statement", "Corporate Author", "Corporate Contributors", "Former owner", "Engraver", "Issuance type", "Shelfmarks"]
df.drop(columns = drop_columns, axis = 1, inplace = True)
df.head()

Identifier  ...                                         Flickr URL
0         206  ...  http://www.flickr.com/photos/britishlibrary/ta...
1         216  ...  http://www.flickr.com/photos/britishlibrary/ta...
2         218  ...  http://www.flickr.com/photos/britishlibrary/ta...
3         472  ...  http://www.flickr.com/photos/britishlibrary/ta...
4         480  ...  http://www.flickr.com/photos/britishlibrary/ta...

[5 rows x 8 columns]

**Changing the index of a dataframe**

In [14]:
df["Identifier"].is_unique

True

Let's replace the existing index with the column "Identifier" using *set_index*

In [15]:
df = df.set_index("Identifier", inplace = True)
df.head()

Place of Publication  ...                                         Flickr URL
Identifier                            ...                                                   
206                           London  ...  http://www.flickr.com/photos/britishlibrary/ta...
216         London; Virtue & Yorston  ...  http://www.flickr.com/photos/britishlibrary/ta...
218                           London  ...  http://www.flickr.com/photos/britishlibrary/ta...
472                           London  ...  http://www.flickr.com/photos/britishlibrary/ta...
480                           London  ...  http://www.flickr.com/photos/britishlibrary/ta...

[5 rows x 7 columns]

In [16]:
df.loc[206]

Place of Publication                                               London
Date of Publication                                           1879 [1878]
Publisher                                                S. Tinsley & Co.
Title                                   Walter Forbes. [A novel.] By A. A
Author                                                              A. A.
Contributors                                              FORBES, Walter.
Flickr URL              http://www.flickr.com/photos/britishlibrary/ta...
Name: 206, dtype: object

**Tidying up Fields in the Data**

In [18]:
df.isnull().sum()

Place of Publication       0
Date of Publication      181
Publisher               4195
Title                      0
Author                  1778
Contributors               0
Flickr URL                 0
dtype: int64

Let's clean the columns "Data of Publication" & "Place of publication"

In [19]:
df.dtypes

Place of Publication    object
Date of Publication     object
Publisher               object
Title                   object
Author                  object
Contributors            object
Flickr URL              object
dtype: object

One field where it makes sense to enforce a numeric value is the "Date of publication" so that we can do calculations down the road:

In [20]:
df.loc[1905:, "Date of Publication"].head(10)

Identifier
1905           1888
1929    1839, 38-54
2836           1897
2854           1865
2956        1860-63
2957           1873
3017           1866
3131           1899
4598           1814
4884           1820
Name: Date of Publication, dtype: object

A particular book can have only one date of publication. Therefore, we need to do the following:

* Remove the extra dates in square brackets, wherever present: 1879 [1878]
* Convert date ranges to their “start date”, wherever present: 1860-63; 1839, 38-54
* Completely remove the dates we are not certain about and replace them with NumPy’s NaN: [1897?]
* Convert the string nan to NumPy’s NaN value

regular expression to extract the publication year

In [21]:
regex = r"^(\d{4})"

In [22]:
extr = df["Date of Publication"].str.extract(regex, expand = False)
extr.head()

Identifier
206    1879
216    1868
218    1869
472    1851
480    1857
Name: Date of Publication, dtype: object

We can get it's easily numerical values

In [25]:
df["Date of Publication"] = pd.to_numeric(extr)
df["Date of Publication"].dtypes

dtype('float64')

In [27]:
df["Date of Publication"].isnull().sum() / len(df)

0.11717147339205986

This results in about one in every 10% values being missing, which is a small price to pay for now being able to do computations on the remaining valid values